OTG Test - RT-1.4: BGP Graceful Restart

octpetre opened the pull request:
OTG Test conversion of **_RT-1.4: BGP Graceful Restart_**
Resolves #580 
The ACL sections were replaced by the "Initiate graceful restart" in the case of DUT helper and "gnoi restart pid" in case of DUT restart initiator

octpetre committed:
ate test copy

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'main' of github.com:openconfig/featureprofiles into otg-rt1.4

octpetre committed:
otg related changes

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
sync to the ATE test

octpetre committed:
Merge branch 'otg-rt1.4' of github.com:open-traffic-generator/featureprofiles into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
-removed ACL
-updated README
-updated the OTG configuration
-added gnoi commands

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
replaced rundata with metadata

octpetre committed:
adding gnoi service to Arista

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
changed gRIBIDaemons map name

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
deviations updates

octpetre committed:
Merge branch 'otg-rt1.4' of github.com:open-traffic-generator/featureprofiles into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' of github.com:open-traffic-generator/featureprofiles into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
updated metadata.textproto

octpetre committed:
added missing deviation

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
corrected to src ip instead of dstIP for routes

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Removed not needed route policies

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
corrected gnoi client init

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

jasdeep-hundal commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,677 @@
+		startTime = time.Now()
+		t.Log("Send Traffic while GR timer counting down. Traffic should pass as BGP GR is enabled!")
+		t.Log("Send Graceful Restart Trigger from OTG to DUT -- trigger GR Restart timer(60s) < configured GR restart Timer(120s)")
+		ate.OTG().SetControlAction(t, createGracefulRestartAction(t, []string{"atedst.BGP4.peer"}, 60))
```
Move 60 to a constant next to grTime so that it's easier to see when both should be changed.

octpetre replied:
Done

jasdeep-hundal commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,677 @@
+		}
+	})
+
+	const stopDuration = 45 * time.Second
```
Should this either be a constant or based on the 60 second value (as a constant?)

octpetre replied:
Done

jasdeep-hundal commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,677 @@
+	})
+
+	// Send Graceful Restart Trigger from ATE to DUT after the expiration of the GR timer configured on the DUT
+	t.Run("InitiateGracefulRestartExceedingDUTGRTimer", func(t *testing.T) {
+		t.Log("Send Graceful Restart Trigger from OTG to DUT -- Trigger GR Restart timer(140s) > configured GR Stale Time (120s) ")
+		ate.OTG().SetControlAction(t, createGracefulRestartAction(t, []string{"atedst.BGP4.peer"}, grStaleRouteTime+20))
+		time.Sleep(time.Second * grStaleRouteTime)
+		ate.OTG().StartTraffic(t)
+	})
```
I don't think we're distinguishing between the GR restart/stale times in this case. (As I understand it, the routes would be dropped from the RIB after the advertised restart time from the ATE anyways.)

@sachendras :
- Is my understanding correct here?
- Is it possible (if arbitrary) to test the stale time setting on the DUT by setting it lower than the GR restart time? Or is that invalid configuration?

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
corrected gosnappi.NewConfig

octpetre commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,677 @@
+
+	// Configure interface on the DUT
+	t.Log("Start DUT interface Config")
+	configureDUT(t, dut)
```
Added

octpetre commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,677 @@
+	nbrList := buildNbrList(ateAS)
+	dutConf := bgpWithNbr(dutAS, nbrList, dut)
+	gnmi.Replace(t, dut, dutConfPath.Config(), dutConf)
+	fptest.LogQuery(t, "DUT BGP Config", dutConfPath.Config(), gnmi.GetConfig(t, dut, dutConfPath.Config()))
```
Done

octpetre commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,677 @@
+		SetRxNames([]string{dstIpv4.Name()})
+	flowipv4.Size().SetFixed(512)
+	flowipv4.Rate().SetPps(100)
+	flowipv4.Duration().SetChoice("continuous")
```
Not sure what you are reffering to, but I removed the "continuous" which was default anyway

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,677 @@
+	dstIpv4 := dstEth.Ipv4Addresses().Items()[0]
+	dstIpv6 := dstEth.Ipv6Addresses().Items()[0]
+
+	srcBgp := srcDev.Bgp().SetRouterId(ateSrc.IPv4)
```
Done

octpetre committed:
PR review changes

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
removed setChoice

MarcCharlebois commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/metadata.textproto:
```diff
@@ -0,0 +1,56 @@
+platform_exceptions: {
+  platform: {
+    vendor: NOKIA
+    hardware_model: "7250-ixr10e"
+    hardware_model: "ixr10"
```
Not required - those lines can be removed

octpetre replied:
removed

MarcCharlebois commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/metadata.textproto:
```diff
@@ -0,0 +1,56 @@
+    explicit_port_speed: true
+    explicit_interface_in_default_vrf: true
+    route_policy_under_afi_unsupported: true
+  }
```
missing `interface_enabled: true`

octpetre replied:
added

MarcCharlebois commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/metadata.textproto:
```diff
@@ -0,0 +1,56 @@
+    hardware_model: "ixr10"
+  }
+  deviations: {
+    use_vendor_native_acl_config: true
```
Deviation not required; this was a limitation of IxNetwork (OTG supports GR natively - no need for ACLs)

octpetre replied:
removed

MarcCharlebois commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,689 @@
+	advertisedRoutesv6Net2   = "2001:db8::203:0:113:3"
+	advertisedRoutesv4Prefix = 32
+	advertisedRoutesv6Prefix = 128
+	aclNullPrefix            = "0.0.0.0/0"
+	aclv6NullPrefix          = "::/0"
+	aclName                  = "BGP-DENY-ACL"
+	aclv6Name                = "ipv6-policy-acl"
```
Probably not required as no ACLs are configured?

octpetre replied:
Indeed...removed

MarcCharlebois commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,689 @@
+	pgv6 := bgp.GetOrCreatePeerGroup(peerv6GrpName)
+	pgv6.PeerAs = ygot.Uint32(ateAS)
+	pgv6.PeerGroupName = ygot.String(peerv6GrpName)
+
```
deviation is set in metadata but not included in testcase.
Ref deviation from ATE test: https://github.com/openconfig/featureprofiles/blob/066ccf7cc3dc4bd797e6e4670a6062a7b0328737/feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go#L199-L220

LimeHat commented:
Note that the readme was updated in #1791 

The PR still includes the previous readme version

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,689 @@
+	var pId uint64
+	pName := BGPDaemons[dut.Vendor()]
+
+	t.Run("FindGRIBIDaemonPid", func(t *testing.T) {
```
corrected

octpetre committed:
addressed pr review comments

octpetre commented:
> Note that the readme was updated in #1791
> 
> The PR still includes the previous readme version

Indeed. However with the new README, who should consider breaking up the test. @jasdeep-hundal what do you think?

MarcCharlebois commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,708 @@
+	statePath := gnmi.OC().NetworkInstance(deviations.DefaultNetworkInstance(dut)).Protocol(oc.PolicyTypes_INSTALL_PROTOCOL_TYPE_BGP, "BGP").Bgp()
+	nbrPath := statePath.Neighbor(ateDst.IPv4)
+
+	t.Run("VerifyBGPNOTEstablished", func(t *testing.T) {
+		t.Log("Waiting for BGP neighbor to go to ACTIVE state after GR Trigger to DUT...")
+		bgpState := oc.Bgp_Neighbor_SessionState_ACTIVE
+		_, ok := gnmi.Watch(t, dut, nbrPath.SessionState().State(), 2*time.Minute, func(val *ygnmi.Value[oc.E_Bgp_Neighbor_SessionState]) bool {
+			currState, ok := val.Val()
+			return ok && currState == bgpState
+		}).Await(t)
+		if !ok {
+			fptest.LogQuery(t, "BGP reported state", nbrPath.State(), gnmi.Get(t, dut, nbrPath.State()))
+			t.Errorf("BGP session did not go Down as expected")
+		}
+	})
```
Should there be a mechanism here preventing the DUT/ATE to re-establish the BGP connection after killing the bgp daemon? It seems like this was included in the [new readme](https://github.com/openconfig/featureprofiles/commit/08c902105b58a58748a665646452976e566ae02a#diff-2ab3ef14592d0bf4292c57211f4ccf8d95041656e417fc2bab51d9f72bad53d9R34-R42). @jasdeep-hundal, thoughts?

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
major rework - initial commit

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
moved the configure ATE before configure DUT

octpetre committed:
Corrected the dst peers

octpetre committed:
better logging and timing

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
updated README

octpetre committed:
Added RT-1.4.6 test

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
-corrections to the README
-corrected the direction of traffic for 1.4.7 -- 1.4.10

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre commented:
@sachendras could you please review the corrections I've made to the README file. I'm assuming we would want to run 1.4.5 for both IBGP and EBGP. Also for tests 1.4.7 through 1.4.10 when testing EBGP I'm assuming the direction of traffic is ATE port2 - ATE Port1 (EBGP routes). In this case the notification will be sent from DUT port1

octpetre committed:
added 1.4.7 - 1.4.10

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

jasdeep-hundal commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,977 @@
+	pgGrV4.Enabled = ygot.Bool(true)
+	pgGrV4.RestartTime = ygot.Uint16(grRestartTime)
+	pgGrV4.StaleRoutesTime = ygot.Uint16(grStaleRouteTime)
+	// pg.PeerAs = ygot.Uint32(ateAS)
```
Can be removed?

octpetre replied:
done

jasdeep-hundal commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,977 @@
+	pgGrV6.Enabled = ygot.Bool(true)
+	pgGrV6.RestartTime = ygot.Uint16(grRestartTime)
+	pgGrV6.StaleRoutesTime = ygot.Uint16(grStaleRouteTime)
+	// pgv6.PeerAs = ygot.Uint32(ateAS)
```
Can be removed?

octpetre replied:
done

jasdeep-hundal commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,977 @@
+	pg := bgp.GetOrCreatePeerGroup(peerv4GrpName)
+	pgGrV4 := pg.GetOrCreateGracefulRestart()
+	pgGrV4.Enabled = ygot.Bool(true)
+	pgGrV4.RestartTime = ygot.Uint16(grRestartTime)
+	pgGrV4.StaleRoutesTime = ygot.Uint16(grStaleRouteTime)
```
These (RestartTime and StaleRoutesTime) are indicated in the README as set at the 'global' level. (I believe the existing ATE test does do that.) Might also imply changes in `checkBgpGRConfig`?

octpetre replied:
Yes, they are defined at the global level and also verified in the checkBgpGRConfig at that level 
```isGrEnabled := gnmi.Get(t, dut, statePath.Global().GracefulRestart().Enabled().State())```

jasdeep-hundal commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
If we do wind up sending traffic from both ATE interfaces, do we want to rename config in general to reflect IBGP vs. EBGP instead? (Eg. srcEth -> iBGPEth maybe).

octpetre replied:
renamed the variables and everything related to reflect EBGP and IBGP settings

octpetre committed:
Merge branch 'main' into otg-rt1.4

MarcCharlebois commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,977 @@
+	t.Log("Configure/update Network Instance")
+	fptest.ConfigureDefaultNetworkInstance(t, dut)
+
+	if deviations.ExplicitPortSpeed(dut) {
```
This deviation can be removed

octpetre replied:
done

octpetre replied:
removed

MarcCharlebois commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/metadata.textproto:
```diff
@@ -0,0 +1,56 @@
+    vendor: NOKIA
+      }
+  deviations: {
+    explicit_port_speed: true
```
Can be removed, please.

octpetre replied:
done

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,977 @@
+	dstBgp6PeerRoutes.Addresses().Add().SetAddress(ibgpV6AdvStartRoute).SetPrefix(advertisedRoutesv6Prefix).SetCount(routeCount)
+
+	ate.OTG().PushConfig(t, config)
+	// ate.OTG().StartProtocols(t)
+	t.Log("Pushing config to ATE and starting protocols...")
+
```
removed

octpetre commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,977 @@
+
+	srcBgp := srcDev.Bgp().SetRouterId(srcIpv4.Address())
+	srcBgp4Peer := srcBgp.Ipv4Interfaces().Add().SetIpv4Name(srcIpv4.Name()).Peers().Add().SetName(ateSrc.Name + ".BGP4.peer")
+	srcBgp4Peer.Advanced().SetKeepAliveInterval(keepaliveTimer).SetHoldTimeInterval(3 * keepaliveTimer)
```
the hold timer is ussually 3 times longer than the keep-alive..which means that after 3 failed keep-alives the session is considered down

octpetre committed:
partial requested changes

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,977 @@
+}
+
+func TestBGPPGracefulRestart(t *testing.T) {
+	cases := []struct {
```
moved them after RT-1.4.1

octpetre committed:
renaming vars and moved test cases block

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
fixes to 1.4.8 and 1.4.10

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
better timing logging
added skipReason
fixed the unblockACL

octpetre committed:
Merge branch 'otg-rt1.4' of github.com:open-traffic-generator/featureprofiles into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
some README updates

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

yini101 commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,981 @@
+			ate := ondatra.ATE(t, "ate")
+			// ATE Configuration.
+			t.Log("Start ATE Config")
+			configureATE(t, ate, 60)
```
Seems like `ConfigureDUT()` is missing. Could you please add it?

octpetre replied:
@praveen-puli Should we add this change where we prepare the DUT before each test variation?

navaneethyv replied:
hard agree , this is a must

octpetre replied:
Code added

yini101 commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,981 @@
+	}, {
+		name:         "RT-1.4.9 Send Hard Notification",
+		direction:    "send",
+		notification: "soft",
```
Test case name says hard, but notification is soft. Please change it to hard.
```suggestion
		notification: "hard",
```

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,981 @@
+		t.Logf("Graceful restart timer enabled as expected to be %v", grRestartTime)
+	} else {
+		t.Errorf("Expected Graceful restart timer: got %v, want %v", grTimerValV6, grRestartTime)
+	}
```
Done

octpetre committed:
requested changes added

yini101 commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,981 @@
+	triggerGrTimer           = 180
+	stopDuration             = 45 * time.Second
+	grRestartTime            = 120
+	grStaleRouteTime         = 120
```
I see, I will raise an issue with them about this. It should be okay if we don't change this for now. Thanks.

octpetre committed:
Merge branch 'openconfig:main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre commented:
Just ran a test in kne using an old Cisco 8000e version. See the results below.

```
--- FAIL: TestBGPPGracefulRestart (2453.53s)
    --- PASS: TestBGPPGracefulRestart/RT-1.4.1_Enable_and_validate_BGP_Graceful_restart_feature (79.45s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.1_Enable_and_validate_BGP_Graceful_restart_feature/verify_DUT_Ports (2.05s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.1_Enable_and_validate_BGP_Graceful_restart_feature/Verify_BGP_Parameters (5.81s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.1_Enable_and_validate_BGP_Graceful_restart_feature/Check_BGP_status (67.53s)
    --- FAIL: TestBGPPGracefulRestart/RT-1.4.2_Restart_DUT_Speaker_Gracefully (51.48s)
    --- FAIL: TestBGPPGracefulRestart/RT-1.4.3_Restart_DUT_Speaker_Abruptly (91.62s)
    --- PASS: TestBGPPGracefulRestart/RT-1.4.4_Restart_Receiver_Gracefully (446.77s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.4_Restart_Receiver_Gracefully/Verify_No_Packet_Loss_for_IBGP (1.06s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.4_Restart_Receiver_Gracefully/Confirm_Packet_Loss_for_IBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.4_Restart_Receiver_Gracefully/Verify_No_Packet_Loss_for_EBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.4_Restart_Receiver_Gracefully/Confirm_Packet_Loss_for_EBGP (1.01s)
    --- FAIL: TestBGPPGracefulRestart/RT-1.4.5_Restart_Receiver_Abruptly (416.82s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.5_Restart_Receiver_Abruptly/Verify_No_Packet_Loss_for_IBGP (1.01s)
        --- FAIL: TestBGPPGracefulRestart/RT-1.4.5_Restart_Receiver_Abruptly/Confirm_Packet_Loss_for_IBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.5_Restart_Receiver_Abruptly/Verify_No_Packet_Loss_for_EBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.5_Restart_Receiver_Abruptly/Confirm_Packet_Loss_for_EBGP (1.01s)
    --- FAIL: TestBGPPGracefulRestart/RT-1.4.6_Test_support_for_RFC8538_compliance_by_letting_Hold-time_expire (503.25s)
        --- FAIL: TestBGPPGracefulRestart/RT-1.4.6_Test_support_for_RFC8538_compliance_by_letting_Hold-time_expire/Verify_No_Packet_Loss_for_IBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.6_Test_support_for_RFC8538_compliance_by_letting_Hold-time_expire/Confirm_Packet_Loss_for_IBGP (1.01s)
        --- FAIL: TestBGPPGracefulRestart/RT-1.4.6_Test_support_for_RFC8538_compliance_by_letting_Hold-time_expire/Verify_No_Packet_Loss_for_EBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.6_Test_support_for_RFC8538_compliance_by_letting_Hold-time_expire/Confirm_Packet_Loss_for_EBGP (1.01s)
    --- SKIP: TestBGPPGracefulRestart/RT-1.4.7_Send_Soft_Notification (0.00s)
    --- FAIL: TestBGPPGracefulRestart/RT-1.4.8_Receive_Soft_Notification (434.05s)
        --- FAIL: TestBGPPGracefulRestart/RT-1.4.8_Receive_Soft_Notification/Verify_No_Packet_Loss_for_EBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.8_Receive_Soft_Notification/Confirm_Packet_Loss_for_EBGP (1.01s)
        --- FAIL: TestBGPPGracefulRestart/RT-1.4.8_Receive_Soft_Notification/Verify_No_Packet_Loss_for_IBGP (1.02s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.8_Receive_Soft_Notification/Confirm_Packet_Loss_for_IBGP (1.01s)
    --- SKIP: TestBGPPGracefulRestart/RT-1.4.9_Send_Hard_Notification (0.00s)
    --- FAIL: TestBGPPGracefulRestart/RT-1.4.10_Receive_Hard_Notification (430.09s)
        --- FAIL: TestBGPPGracefulRestart/RT-1.4.10_Receive_Hard_Notification/Verify_No_Packet_Loss_for_EBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.10_Receive_Hard_Notification/Confirm_Packet_Loss_for_EBGP (1.01s)
        --- FAIL: TestBGPPGracefulRestart/RT-1.4.10_Receive_Hard_Notification/Verify_No_Packet_Loss_for_IBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.10_Receive_Hard_Notification/Confirm_Packet_Loss_for_IBGP (1.01s)
FAIL
*** PROPERTY: validator.core.end -> 
Delta Core Files by DUT: 
DUT: dut

*** Releasing the testbed...

*** PROPERTY: time.begin -> 1785267012
*** PROPERTY: time.end -> 1785269510
FAIL    command-line-arguments  2499.817s
FAIL
```

octpetre commented:
Also with ceos 

```
--- FAIL: TestBGPPGracefulRestart (2001.70s)
    --- PASS: TestBGPPGracefulRestart/RT-1.4.1_Enable_and_validate_BGP_Graceful_restart_feature (6.97s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.1_Enable_and_validate_BGP_Graceful_restart_feature/verify_DUT_Ports (0.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.1_Enable_and_validate_BGP_Graceful_restart_feature/Verify_BGP_Parameters (0.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.1_Enable_and_validate_BGP_Graceful_restart_feature/Check_BGP_status (4.00s)
    --- FAIL: TestBGPPGracefulRestart/RT-1.4.2_Restart_DUT_Speaker_Gracefully (15.77s)
    --- FAIL: TestBGPPGracefulRestart/RT-1.4.3_Restart_DUT_Speaker_Abruptly (15.29s)
    --- PASS: TestBGPPGracefulRestart/RT-1.4.4_Restart_Receiver_Gracefully (348.38s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.4_Restart_Receiver_Gracefully/Verify_No_Packet_Loss_for_IBGP (1.04s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.4_Restart_Receiver_Gracefully/Confirm_Packet_Loss_for_IBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.4_Restart_Receiver_Gracefully/Verify_No_Packet_Loss_for_EBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.4_Restart_Receiver_Gracefully/Confirm_Packet_Loss_for_EBGP (1.01s)
    --- PASS: TestBGPPGracefulRestart/RT-1.4.5_Restart_Receiver_Abruptly (349.36s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.5_Restart_Receiver_Abruptly/Verify_No_Packet_Loss_for_IBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.5_Restart_Receiver_Abruptly/Confirm_Packet_Loss_for_IBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.5_Restart_Receiver_Abruptly/Verify_No_Packet_Loss_for_EBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.5_Restart_Receiver_Abruptly/Confirm_Packet_Loss_for_EBGP (1.01s)
    --- FAIL: TestBGPPGracefulRestart/RT-1.4.6_Test_support_for_RFC8538_compliance_by_letting_Hold-time_expire (542.62s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.6_Test_support_for_RFC8538_compliance_by_letting_Hold-time_expire/Verify_No_Packet_Loss_for_IBGP (1.02s)
        --- FAIL: TestBGPPGracefulRestart/RT-1.4.6_Test_support_for_RFC8538_compliance_by_letting_Hold-time_expire/Confirm_Packet_Loss_for_IBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.6_Test_support_for_RFC8538_compliance_by_letting_Hold-time_expire/Verify_No_Packet_Loss_for_EBGP (1.01s)
        --- FAIL: TestBGPPGracefulRestart/RT-1.4.6_Test_support_for_RFC8538_compliance_by_letting_Hold-time_expire/Confirm_Packet_Loss_for_EBGP (1.01s)
    --- SKIP: TestBGPPGracefulRestart/RT-1.4.7_Send_Soft_Notification (0.00s)
    --- PASS: TestBGPPGracefulRestart/RT-1.4.8_Receive_Soft_Notification (360.57s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.8_Receive_Soft_Notification/Verify_No_Packet_Loss_for_EBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.8_Receive_Soft_Notification/Confirm_Packet_Loss_for_EBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.8_Receive_Soft_Notification/Verify_No_Packet_Loss_for_IBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.8_Receive_Soft_Notification/Confirm_Packet_Loss_for_IBGP (1.01s)
    --- SKIP: TestBGPPGracefulRestart/RT-1.4.9_Send_Hard_Notification (0.00s)
    --- FAIL: TestBGPPGracefulRestart/RT-1.4.10_Receive_Hard_Notification (362.72s)
        --- FAIL: TestBGPPGracefulRestart/RT-1.4.10_Receive_Hard_Notification/Verify_No_Packet_Loss_for_EBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.10_Receive_Hard_Notification/Confirm_Packet_Loss_for_EBGP (1.01s)
        --- FAIL: TestBGPPGracefulRestart/RT-1.4.10_Receive_Hard_Notification/Verify_No_Packet_Loss_for_IBGP (1.01s)
        --- PASS: TestBGPPGracefulRestart/RT-1.4.10_Receive_Hard_Notification/Confirm_Packet_Loss_for_IBGP (1.01s)
FAIL
*** PROPERTY: validator.core.end -> 
Delta Core Files by DUT:

*** Releasing the testbed...

*** PROPERTY: time.begin -> 1785279727
*** PROPERTY: time.end -> 1785281741
FAIL	command-line-arguments	2016.587s
FAIL
```

navaneethyv commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,972 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
```
Please rebase it so that diff can be viewed against

https://github.com/openconfig/featureprofiles/blob/main/feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go

octpetre replied:
the OTG version is completely reworked based on the latest README

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
added requested code

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,981 @@
+
+func unblockBGPTCP(t *testing.T, iface *oc.Acl_Interface, dutIfName string) {
+	dut := ondatra.DUT(t, "dut")
+	aclPath := configACLInterface(iface, dutIfName)
```
done

octpetre committed:
Merge branch 'main' into otg-rt1.4

octpetre committed:
Merge branch 'main' into otg-rt1.4

AmrNJ reviewed (changes requested):
The existing test needs to be modified instead of creating a new test.

octpetre commented:
> The existing test needs to be modified instead of creating a new test.

The old test was completely different as the README went through many iteration changes. So I removed the old test. Please let me know if the path is good featureprofiles/feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go

AmrNJ commented:
/gemini review

AmrNJ commented:
This is a very old PR, can you create a new one? 
I see some mismatch between the old and new test. The new code shouldn't deprecate interms functionality and coverage. 
- Missing `bgpgr.HelperOnly = ygot.Bool(false)`
- Old code had GR enabled ONLY on the peer-group, and once with on the neighbor level. Enables GR at the Global, Peer-Group, and Neighbor levels all at the same time.
- AFI-SAFI GR state path check

octpetre commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,986 @@
+	v4 := flowipv4.Packet().Add().Ipv4()
+	v4.Src().SetValue(src.IPv4)
+	v4.Dst().Increment().SetStart(dstStart).SetCount(routeCount)
+
+	ate.OTG().PushConfig(t, config)
+	ate.OTG().StartProtocols(t)
+
```
added

octpetre commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,986 @@
+				t.Logf("Waiting for %s short of stale route time expiration of %v", waitDuration, grStaleRouteTime)
+				time.Sleep(waitDuration)
+				ate.OTG().StopTraffic(t)
+				t.Run("Verify No Packet Loss for "+mode, func(t *testing.T) {
+					verifyNoPacketLoss(t, ate)
+				})
```
added

octpetre commented on feature/bgp/gracefulrestart/otg_tests/bgp_graceful_restart_test/bgp_graceful_restart_test.go:
```diff
@@ -0,0 +1,986 @@
+				}
+
+				sendTraffic(t, ate)
+				t.Run("Confirm Packet Loss for "+mode, func(t *testing.T) {
+					confirmPacketLoss(t, ate)
+				})
```
added

octpetre commented:
> * Enables GR at the Global, Peer-Group, and Neighbor levels all at the same time.

- I'm not sure about the missing bgpgr.HelperOnly = ygot.Bool(false).
- I've added the GR at the neighbor level...I changed the README as well to say that.
- for AFI/SAFI GR state path check....do I need to enable GR at that level too?

octpetre commented:
As requested...opening a new PR
